### ISSUE TITLE
Improvements to doc testing

### ISSUE THREAD

**Issue by tengqm** [kubernetes/website#22348]
**This is a Feature Request**

**What would you like to be added**

Improvements to validating/verifying sample manifests.

**Why is this needed**

We have many sample manifests under the `content/<lang>/examples` directories. They are extracted from the markdown sources and then consolidated into the `../examples` directory so that they can be easily referenced, reused, and copy-pasted by users.

We have tools (`content/<lang>/examples/examples_test.go`) for automating the validation of these manifests against the *current* Kubernetes release. This validation is important given that features are constantly added/deprecated in the upstream project. We want to ensure that what we present on the website always work with *current* Kubernetes release.

TLDR:
When the testing tool was introduced, there wasn't an easy way to vendor Kubernetes code for testing. We had to manipulate the project layout after cloning the kubernetes/kubernetes project due to some constraints in the way Golang searches for modules. This is no longer a big problem when the community converges to go-mod dependency management. With changes made in #22011, we can test the sample manifests easily:

```
go test k8s.io/website/content/<lang>/examples
```

However, we still see rooms for improvements, namely:

- [ ] Make the testing code a standalone tool rather than a Go test case, so we can better control the output from the tool.
- [ ] Discover the resources from the YAML/JSON files automatically by searching for root-level `kind` keys, so that
    * we don't need to hardcode the resource types to check in the Go source
    * future changes to the `examples/` directory do not mandate a change the testing tools
- [ ] Make the testing tool part of the gate, and ideally do *NOT* run the tool if a PR is not touching any sample manifests.
   (we used to have this for travis CI but not for netlify).
- [ ] Make the testing tool a binary, e.g. a container image, that can be source from an external repo (e.g. a kubernetes-sigs project).

**Comments**

There are other tools that can be consolidated, such "link checkers", "content-type checkers", etc.

**Comment by kbhawkey** [comment]
Hi! I think it would be good to prioritize this work. What do you think?
Could GitHub Actions trigger the testing of any changed example files?

@tengqm ,  Can you provide more information about:
`Discover the resources from the YAML/JSON files automatically by searching for root-level kind keys, so that we don't need to hardcode the resource types to check in the Go source`?

The area should be testing not web-development?

**Comment by tengqm** [comment]
> @tengqm , Can you provide more information about:
> Discover the resources from the YAML/JSON files automatically by searching for root-level kind keys, so that we don't need > to hardcode the resource types to check in the Go source?

By resource discovery, I meant a replacement of the huge map [here](https://github.com/kubernetes/website/blob/master/content/en/examples/examples_test.go#L370-L660) which obviously is not an elegant solution going forward.
